File: java/linkedlist/PartitionList.java
/**
 * Q13. Partition List
 * 
 * Problem:
 * Given the head of a linked list and a value x, partition it such that all nodes less than x
 * come before nodes greater than or equal to x. You should preserve the original relative order
 * of the nodes in each of the two partitions.
 * 
 * Example:
 * Input: head = [1,4,3,2,5,2], x = 3
 * Output: [1,2,2,4,3,5]
 * 
 * Approach:
 * 1. Create two dummy nodes for before and after lists
 * 2. Use two pointers to build before and after lists
 * 3. Traverse original list and partition nodes
 * 4. Connect before and after lists
 * 5. Return the merged list
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class PartitionList {
    public ListNode partition(ListNode head, int x) {
        // Create dummy nodes for before and after lists
        ListNode beforeHead = new ListNode(0);
        ListNode afterHead = new ListNode(0);
        ListNode before = beforeHead;
        ListNode after = afterHead;

        // Traverse the list and partition nodes
        while (head != null) {
            if (head.val < x) {
                before.next = head;
                before = before.next;
            } else {
                after.next = head;
                after = after.next;
            }
            head = head.next;
        }

        // Connect before and after lists
        after.next = null;
        before.next = afterHead.next;

        return beforeHead.next;
    }

    // Main method for testing
    public static void main(String[] args) {
        // Create test linked list: 1->4->3->2->5->2
        ListNode head = new ListNode(1);
        head.next = new ListNode(4);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(2);
        head.next.next.next.next = new ListNode(5);
        head.next.next.next.next.next = new ListNode(2);

        // Partition list around x = 3
        PartitionList solution = new PartitionList();
        ListNode result = solution.partition(head, 3);

        // Print the result
        System.out.print("List after partitioning around 3: ");
        while (result != null) {
            System.out.print(result.val + " -> ");
            result = result.next;
        }
        System.out.println("null");

        // Test with all elements less than x
        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(2);
        head2.next.next = new ListNode(3);

        ListNode result2 = solution.partition(head2, 4);
        System.out.print("List with all elements < 4: ");
        while (result2 != null) {
            System.out.print(result2.val + " -> ");
            result2 = result2.next;
        }
        System.out.println("null");

        // Test with all elements greater than x
        ListNode head3 = new ListNode(5);
        head3.next = new ListNode(6);
        head3.next.next = new ListNode(7);

        ListNode result3 = solution.partition(head3, 4);
        System.out.print("List with all elements > 4: ");
        while (result3 != null) {
            System.out.print(result3.val + " -> ");
            result3 = result3.next;
        }
        System.out.println("null");
    }
} 

File: javascript/linkedlist/PartitionList.js
/**
 * Partition List
 * 
 * Problem:
 * Given the head of a linked list and a value x, partition it such that all nodes
 * less than x come before nodes greater than or equal to x. You should preserve
 * the original relative order of the nodes in each of the two partitions.
 * 
 * Example:
 * Input: head = [1,4,3,2,5,2], x = 3
 * Output: [1,2,2,4,3,5]
 * Explanation:
 * - Nodes less than 3: [1,2,2]
 * - Nodes greater than or equal to 3: [4,3,5]
 * - Preserved relative order in both partitions
 * 
 * Approach:
 * 1. Create two dummy nodes for less and greater lists
 * 2. Traverse the original list:
 *    - If node value < x, add to less list
 *    - If node value >= x, add to greater list
 * 3. Connect the two lists
 * 4. Return the head of the combined list
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class PartitionList {
    /**
     * @param {ListNode} head
     * @param {number} x
     * @return {ListNode}
     */
    partition(head, x) {
        if (!head || !head.next) return head;

        // Create dummy nodes for less and greater lists
        const lessDummy = new ListNode(0);
        const greaterDummy = new ListNode(0);
        let less = lessDummy;
        let greater = greaterDummy;
        let current = head;

        // Partition the list
        while (current) {
            if (current.val < x) {
                less.next = current;
                less = less.next;
            } else {
                greater.next = current;
                greater = greater.next;
            }
            current = current.next;
        }

        // Connect the two lists
        greater.next = null;
        less.next = greaterDummy.next;

        return lessDummy.next;
    }

    // Helper method to create a linked list
    static createList(values) {
        if (!values || values.length === 0) return null;
        
        const head = new ListNode(values[0]);
        let current = head;
        
        for (let i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }
        
        return head;
    }

    // Helper method to print the list
    static printList(head) {
        let current = head;
        const values = [];
        while (current) {
            values.push(current.val);
            current = current.next;
        }
        console.log(values.join(' -> '));
    }
}

// Test cases
function runTests() {
    const solution = new PartitionList();

    // Test case 1: Regular case
    console.log("Test case 1: Regular case");
    const list1 = PartitionList.createList([1, 4, 3, 2, 5, 2]);
    console.log("Original list:");
    PartitionList.printList(list1);
    console.log("List after partitioning (x=3):");
    PartitionList.printList(solution.partition(list1, 3));

    // Test case 2: All elements less than x
    console.log("\nTest case 2: All elements less than x");
    const list2 = PartitionList.createList([1, 2, 3, 4, 5]);
    console.log("Original list:");
    PartitionList.printList(list2);
    console.log("List after partitioning (x=6):");
    PartitionList.printList(solution.partition(list2, 6));

    // Test case 3: All elements greater than x
    console.log("\nTest case 3: All elements greater than x");
    const list3 = PartitionList.createList([1, 2, 3, 4, 5]);
    console.log("Original list:");
    PartitionList.printList(list3);
    console.log("List after partitioning (x=0):");
    PartitionList.printList(solution.partition(list3, 0));

    // Test case 4: Empty list
    console.log("\nTest case 4: Empty list");
    const list4 = null;
    console.log("Original list: null");
    console.log("List after partitioning (x=3):");
    PartitionList.printList(solution.partition(list4, 3));

    // Test case 5: Single node list
    console.log("\nTest case 5: Single node list");
    const list5 = PartitionList.createList([1]);
    console.log("Original list:");
    PartitionList.printList(list5);
    console.log("List after partitioning (x=2):");
    PartitionList.printList(solution.partition(list5, 2));

    // Test case 6: List with duplicate values
    console.log("\nTest case 6: List with duplicate values");
    const list6 = PartitionList.createList([3, 1, 3, 2, 3, 4, 3]);
    console.log("Original list:");
    PartitionList.printList(list6);
    console.log("List after partitioning (x=3):");
    PartitionList.printList(solution.partition(list6, 3));

    // Test case 7: List with negative numbers
    console.log("\nTest case 7: List with negative numbers");
    const list7 = PartitionList.createList([-1, -4, -3, -2, -5, -2]);
    console.log("Original list:");
    PartitionList.printList(list7);
    console.log("List after partitioning (x=-3):");
    PartitionList.printList(solution.partition(list7, -3));
}

// Run the tests
runTests(); 